utils/merkle_proof/src/lib.rs

```diff
@@ -133,67 +153,167 @@ fn calculate_merkle_root(
             root = hash(&input);
         }
     }
-
-    H256::from_slice(&root)
+    Ok(H256::from_slice(&root))
+    // return H256::from_slice(&root);
 }
 
-fn concat(mut vec1: Vec<u8>, mut vec2: Vec<u8>) -> Vec<u8> {
-    vec1.append(&mut vec2);
-    return vec1;
+fn verify_merkle_multiproof(leaves: &[H256],  proof: &[H256], indices: &[usize], root: H256) -> Result<bool, MerkleProofError> {
+    match calculate_multi_merkle_root(leaves, proof, indices) {
+        Ok(calculated_root) => Ok(calculated_root == root),
+        Err(err) => Err(err)
+    }
 }
 
-fn get_generalized_index_bit(index: usize, i: usize) -> bool {
-    ((index >> i) & 0x01) == 1
-}
+fn calculate_multi_merkle_root(leaves: &[H256], proof: &[H256], indices: &[usize]) -> Result<H256, MerkleProofError> {
+    let mut btree_first = HashMap::new();
+    let mut btree_second = HashMap::new();
 
+    if leaves.len() != indices.len() {
+        return Err(MerkleProofError::InvalidParamLength {
+            len_first: leaves.len(), 
+            len_second: indices.len()
+        });
+    }
 
-fn hash_and_concat(h1: H256, h2: H256) -> H256 {
-    H256::from_slice(&hash(&concat(
-        h1.as_bytes().to_vec(),
-        h2.as_bytes().to_vec(),
-    )))
-}
+    let helper_indices = get_helper_indices(indices);
 
-fn get_next_power_of_two(depth: usize) -> usize {
-    2usize.pow(depth as u32)      
-}
+    let mut repetition_counter: usize = 0;
+    for i in helper_indices.iter() {
+        if indices.contains(i) {
+            repetition_counter += 1;
+        }
+    }
 
-//paklausti ar reik:
-// kokiu funkciju man reik?-nera dokumentacijoj;
-// verify_merkle_multiproof? yra dokumentacijoj
+    if indices.len() == 1 {
+        repetition_counter += 1;
+    }
+
+    if (helper_indices.len()-repetition_counter) != proof.len() {
+        return Err(MerkleProofError::InvalidParamLength {
+            len_first: helper_indices.len(), 
+            len_second: proof.len()
+        });
+    }
+
+    for (index, leave) in indices.iter().zip(leaves.iter()) {
+        btree_first.insert(*index, *leave);
+    }
+
+    for (helper_step, proof_step) in helper_indices.iter().zip(proof.iter()) {
+        btree_second.insert(*helper_step, *proof_step);
+    }
+
+    btree_first.extend(btree_second);
+
+    let mut keys: Vec<usize> = vec![];
+
+    for(key, _value) in btree_first.iter_mut() {
+        keys.push(key.clone());
+    }
+
+    keys.sort();
+    keys = reverse_vector(keys);
+
+    let mut biggest: usize = *keys.get(0usize).clone().unwrap();
+    while biggest > 0 {
+        if !keys.contains(&biggest) {
+            keys.push(biggest);
+        }
+        biggest -=1;
+    }
+    keys.sort();
+    keys = reverse_vector(keys);
+    let mut position = 1usize;
+
+    while position < keys.len() {
+        let k = keys.get(position).unwrap();
```

Comment:
Same question as up above. If you are using unwrap not in the tests there is a convention to leave a comment why you think its safe